cpp/include/cudf/detail/utilities/assert.cuh

```diff
@@ -35,3 +35,26 @@
 #else
 #define cudf_assert(e) (static_cast<void>(0))
 #endif
+
+/**
+ * @brief Macro indicating that a location in the code is unreachable.
+ *
+ * The CUDF_UNREACHABLE macro should only be used where CUDF_FAIL cannot be used
+ * due to performance or due to being used in device code. In the majority of
+ * host code situations, an exception should be thrown in "unreachable" code
+ * paths as those usually aren't tight inner loops like they are in device code.
+ *
+ * One example where this macro may be used is in conjunction with dispatchers
+ * to indicate that a function does not need to return a default value because
+ * it has already exhausted all possible cases in a `switch` statement.
+ *
+ * Example usage:
+ * ```
+ * CUDF_UNREACHABLE("Invalid type_id.");
+ * ```
+ */
+#define CUDF_UNREACHABLE(msg)             \
+  do {                                    \
+    assert(false && "Unreachable: " msg); \
```

Comment:
What is the use of the assert here? With it, I'm confused about the purpose of the CUDF_UNREACHABLE marco. Is it meant to be used in truly unreachable code, or something that can be reached if users don't follow the API contract?

Reply:
My understanding is that user errors in calling the API should always be caught in host code, unless they're triggering something that we explicitly state as UB, e.g. an error that would require data introspection. `CUDF_UNREACHABLE` (and previously, `cudf_assert`) are typically to catch developer errors triggering code paths in device code that _should_ be unreachable. Branches of code that use `CUDF_UNREACHABLE` can be optimized out entirely in Release builds since both `assert` and `__builtin_unreachable`  will be compiled out, but in Debug builds the assertion will persist to enable debugging.

Reply:
> but in Debug builds the assertion will persist to enable debugging.
This is the part I wasn't clear on 👍 
Maybe some additional docs are required here or in the PR description.

Reply:
Maybe a link to the [`assert` docs on cppreference](https://en.cppreference.com/w/cpp/error/assert) (or maybe even the [C reference](https://en.cppreference.com/w/c/error/assert) since it's originally a C macro) would be helpful? Perhaps knowledge of how assert works with/without `NDEBUG` defined is something worth socializing.

Reply:
It's not that I didn't know that `assert` only exists in debug build, the intent was not clear to me even knowing that. I guess this matters less after code review.

Reply:
Ah I see, I didn't realize that was the point of confusion. Yeah maybe worth a small comment like "assert in debug mode to help debug functions that reach supposedly unreachable code".

Reply:
I added a comment in 8c2b1dcb69612efe573028361a8e1d771c49583d to explain the purpose of the assert. Let me know if additional info would be helpful!